src/python/pants/backend/python/lint/pylint/plugin_target_type.py

```diff
@@ -0,0 +1,63 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pants.backend.python.target_types import COMMON_PYTHON_FIELDS, PythonSources
+from pants.engine.target import Dependencies, Target
+from pants.util.ordered_set import FrozenOrderedSet
+
+
+class PylintPluginSources(PythonSources):
+    required = True
+    expected_file_extensions = (".py",)
+
+
+class PylintPluginDependencies(Dependencies):
+    """Addresses to other targets that this plugin depends on.
+
+    These targets must either be third-party Python dependencies
+    (https://pants.readme.io/docs/python-third-party-dependencies) or be located within this
+    target's same directory or a subdirectory.
+    """
+
+
+class PylintSourcePlugin(Target):
+    """A Pylint plugin loaded through source code.
+
+    To load a source plugin:
+
+        1. Write your plugin. See http://pylint.pycqa.org/en/latest/how_tos/plugins.html.
+        2. Define a `pylint_source_plugin` target with the plugin's Python files included in the
+            `sources` field.
+        3. Add the parent directory of your target to the `root_patterns` option in the `[source]`
+            scope. For example, if your plugin is at `build-support/pylint/custom_plugin.py`, add
+            'build-support/pylint'. This is necessary for Pants to know how to tell Pylint to
+            discover your plugin. See https://pants.readme.io/docs/source-roots.
+        4. Add `load-plugins=$module_name` to your Pylint config file. For example, if your Python
+            file is called `custom_plugin.py`, set `load-plugins=custom_plugin`. Set the `config`
+            option in the `[pylint]` scope to point to your Pylint config file.
+        5. Set the option `source_plugins` in the `[pylint]` scope to include this target's
+            address, e.g. `source_plugins = ["build-support/pylint:plugin"]`.
+
+    To instead load a third-party plugin, set the option `extra_requirements` in the `[pylint]`
+    scope (see https://pants.readme.io/docs/python-linters-and-formatters). Set `load-plugins` in
+    your config file, like you'd do with a source plugin.
```

Comment:
This difference between a source and 3rdparty plugin is probably a good reason to have a plugin be defined by one of the existing target types rather than a new target type. That way, someone could add the existing `python_library` or `python_requirement` targets to a single `plugin_targets` list (like the `source_plugins` list here).

Reply:
This presumes that we want targets for 3rd party plugins, compared to right now simply putting a requirement-style string like `flake8-django==3.0` in `--flake8-extra-requirements`. 

I know we use targets with JVM, but I don't think we want that here. There is much less boilerplate with simply putting the string. 

It's also essential that we can keep including our own extra requirements like this:

https://github.com/pantsbuild/pants/blob/6d3d629c5c261cfc47565e402da655f5e1bd1882/src/python/pants/backend/python/lint/isort/subsystem.py#L11

It's easier for us to express that, and for users to override it, via a string. Keep it simple.

Reply:
> I know we use targets with JVM, but I don't think we want that here. There is much less boilerplate with simply putting the string.

For 3rdparty targets, yes. But not for source targets... it means a new target type with its own docs and etc. If you're able to remove the new target type so that this could just be a `python_library`, then that might tip the scales?

Reply:
> But not for source targets... it means a new target type with its own docs and etc. 

Well, we need the docs no matter what. To use a source plugin with Pylint has some weird quirks that must be explained, particularly needing to set up your `source root`. 

The docs could live in these 3 places:

1) The help message for `--pylint-source-plugins`.
2) The website.
3) A new target type.

Downside of #1 is that we lose all newlines and formatting due to how `help` renders. Because this explanation is quite involved, that matters.

Downside of #2 is that this is a really, really niche feature. I don't think we want this taking up real estate on the website, beyond a tiny tip box saying "Run foo for instructions."

--

The other downside of using `python_library` is that we can't override the description of `dependencies` or change `sources` to be required and have no default globs.

Reply:
I agree that putting the docs somewhere is a good idea. But part of the reason that the docs are as large as they are is _because_ this acts a lot like a special case... if it's possible to remove more of that, it's possible that it would be small enough for the option help string.

A bit of a tangent, but: a fourth option for where the docs might live would be to put this in the docstring for the `Subsystem`, and then render more of that on the docsite. Currently https://www.pantsbuild.org/options_reference.html renders only the first line of the `Subsystem`'s docstring (if it exists)... ditto the CLI I think. Rendering all of it would... probably be generally helpful?

> The other downside of using python_library is that we can't override the description of dependencies or change sources to be required and have no default globs.

What is the reason for the dependencies restriction?

Reply:
> What is the reason for the dependencies restriction? Ideally, removing both restrictions would make this less special-casey, and easier to explain.

This is how Pylint works. I spent a lot of time playing around in a project using Pylint directory (i.e. no Pants), and found some restrictions:

* You must use `load-plugins=my_plugin`, not `load_plugins=path.to.my_plugin`
* To get source dependencies working, my original approach of changing `init-hook` to modify `sys.path` failed. For some reason, nothing would be discoverable. We have to instead set `PYTHONPATH`.
* For some reason, source dependencies must be either in the same directory as the plugin or in a subdirectory. They can't refer to other files like `src/python/f.py`, even if Pylint knows how to run on `src/python/f.py`.

I agree it would be great to remove all these restrictions and documentation, but this is what it takes to work with Pylint.

Reply:
> They can't refer to other files like src/python/f.py, even if Pylint knows how to run on src/python/f.py.

With the sourceroot stripped, that would be at just `f.py`, right? That feels like it should allow it to be loaded normally via the PYTHONPATH.

Reply:
Indeed, it does feel like that. But it still fails from my testing. The import must either be in the same directory or a subdirectory.

For example, if we go with your (good) suggestion of putting plugin contents into a `_plugins` directory, then `_plugins/plugin.py` with `PYTHONPATH=./_plugins` would not be able to import `f.py`. It could only import things in `_plugins`.